core/tasks.rs

```diff
@@ -0,0 +1,193 @@
+use futures::task::AtomicWaker;
+use std::marker::PhantomData;
+use std::ops::DerefMut;
+use std::sync::atomic::AtomicBool;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::task::Context;
+use std::task::Poll;
+
+type UnsendTask = Box<dyn FnOnce(&mut v8::HandleScope) + 'static>;
+type SendTask = Box<dyn FnOnce(&mut v8::HandleScope) + Send + 'static>;
+
+static_assertions::assert_not_impl_any!(V8TaskSpawnerFactory: Send);
+static_assertions::assert_not_impl_any!(V8TaskSpawner: Send);
+static_assertions::assert_impl_all!(V8CrossThreadTaskSpawner: Send);
+
+#[derive(Default)]
+pub(crate) struct V8TaskSpawnerFactory {
+  // TODO(mmastrac): ideally we wouldn't box if we could use arena allocation and a max submission size
+  // TODO(mmastrac): we may want to split the Send and !Send tasks
+  /// The set of tasks, non-empty if `has_tasks` is set.
+  tasks: Mutex<Vec<SendTask>>,
+  /// A flag we can poll without any locks.
+  has_tasks: AtomicBool,
+  /// The polled waker, woken on task submission.
+  waker: AtomicWaker,
+  /// This cannot be Send because it may contain `!Send` tasks submitted by a [`V8TaskSpawner`]. It is
+  /// only safe to send this object to another thread if you plan on submitting [`Send`] tasks to it,
+  /// which is what [`V8CrossThreadTaskSpawner`] does.
+  _unsend_marker: PhantomData<*const ()>,
+}
+
+impl V8TaskSpawnerFactory {
+  pub fn new_same_thread_spawner(self: Arc<Self>) -> V8TaskSpawner {
+    V8TaskSpawner {
+      tasks: self,
+      _unsend_marker: PhantomData,
+    }
+  }
+
+  pub fn new_cross_thread_spawner(self: Arc<Self>) -> V8CrossThreadTaskSpawner {
+    V8CrossThreadTaskSpawner { tasks: self }
+  }
+
+  pub fn has_pending_tasks(&self) -> bool {
+    self.has_tasks.load(std::sync::atomic::Ordering::SeqCst)
+  }
+
+  /// Poll this set of tasks, returning a non-empty set of tasks if there have
+  /// been any queued, or registering the waker if not.
+  pub fn poll_inner(&self, cx: &mut Context) -> Poll<Vec<UnsendTask>> {
+    // Check the flag first -- if it's false we definitely have no tasks
+    if !self
+      .has_tasks
+      .swap(false, std::sync::atomic::Ordering::SeqCst)
```

Comment:
What would happen if a task is scheduled after this `poll_inner` atomic check? `has_pending_task` would spin the event loop?

Reply:
Let me look closer at it again but there should only be the following cases as we're performing an atomic swap to false here:

1. has_tasks is true, we have tasks -- they are returned and has_tasks is set to false
2. has_tasks is true, but we lost a race and have no tasks -- we return pending and has_tasks becomes false. The waker is registered.
3. has_tasks is false -- we return pending and register our waker. If we are racing a cross-thread submission, the waker will be woken and we'll be awoken on the next time through

In all cases, has_tasks ends up as false unless a new submission is made. There should be no way for has_tasks to get set to false without tasks being extracted and run, and there should be no way for has_tasks to remain true while we have no tasks (causing a spin).

Reply:
I added some tests for serial and parallel spawning in a multi-threaded runtime to confirm that this cannot deadlock, even under Miri.

Reply:
Sounds good, thanks